### PR TITLE
DP 8 SDPA support for MLA Decode

### PR DESCRIPTION
**Pull request by avoraTT**
### Ticket
- #23022 
- #25129

### Problem description
In the 1xTG case, we cannot fit a replicated cache. As such, the MLA module needs to support DP around the SDPA op. Currently, we support TP=8 in MLA, but this needs to be extended with DP support.

### What's changed

There are 3 main CCL points that are introduced when supporting DP=8 SDPA inside of the current TP=8 MLA decode module:
- Reduce-Scatter before updating cache
- All-to-All before SDPA
- All-to-All after SDPA

For all of these, we do **not** have full CCL support. Here are the workarounds currently employed:

- Reduce-Scatter
    - Since we do not support sub-tile shapes in reduce scatter, we can do combination of permute + pad + slice alongside the existing reduce-scatter to simulate our use case
- All-toAll
    - For functionality, we've reverted to using an All-Gather + Reduce-Scatter + Scaling methodology. Using this, we can simulate an all-to-all

### Checklist
- [x] [TG Deepseek](https://github.com/tenstorrent/tt-metal/actions/runs/16418927683) CI passes

### CODE REVIEW

File: models/demos/deepseek_v3/tt/mla_1d.py [hunk 1]
```diff
@@ -553,8 +554,9 @@ def decode_model_config(
             packer_l1_acc=False,
         )
 
-        q_num_cores = MLA1D.MAX_BATCH_SIZE  # TODO: How to use non-padded batch size here? (might need to be dynamic)
-        block_height = nearest_y((MLA1D.MAX_BATCH_SIZE * num_heads_local) // q_num_cores, ttnn.TILE_SIZE)
+        q_num_cores = num_cores
+        q_num_cores = min(MLA1D.MAX_BATCH_SIZE // mesh_shape[1] * num_heads, q_num_cores)
+        block_height = nearest_y((MLA1D.MAX_BATCH_SIZE // mesh_shape[1] * num_heads) // q_num_cores, ttnn.TILE_SIZE)
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/mla_1d.py]: Please use even_int_div wherever you do integer division to assert it's not rounded down (unless the code does actually work with rounding)

File: models/demos/deepseek_v3/tests/test_mla_1d.py [hunk 2]
```diff
@@ -130,7 +130,7 @@ def test_forward_pass(
         model_config = MLA1D.decode_model_config(hf_config, mesh_row, ccl)
 
     # Create a new model state
-    model_state = MLA1D.create_state(hf_config, mesh_device=mesh_row)
+    model_state = MLA1D.create_state(hf_config, mesh_device=mesh_row, mode=mode)
```
Comment (barci2) [inline models/demos/deepseek_v3/tests/test_mla_1d.py]: the state is supposed to be shared between the modes

Comment (avoraTT) [reply models/demos/deepseek_v3/tests/test_mla_1d.py]: Yes this will temporarily be used because prefill currently does not support DP=8. As soon as we add support for that, this flag will be removed!

File: models/demos/deepseek_v3/tt/mla_1d.py [hunk 3]
```diff
@@ -619,6 +621,28 @@ def decode_model_config(
             topology=ttnn.Topology.Linear,
         )
 
+        # Q all-to-all
+        config["wq_a2a_ag"] = AllGatherAsyncConfig(
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/mla_1d.py:627]: nitpick (not necessary): I personally think in-place dictionary construction looks cleaner than building it up piece by piece (like in MLP1D)

File: models/demos/deepseek_v3/tt/mla_1d.py [hunk 4]
```diff
@@ -668,9 +726,11 @@ def create_state(
         kvpe_cache_layout = ttnn.TILE_LAYOUT
         kvpe_cache_mem_config = ttnn.DRAM_MEMORY_CONFIG
 
+        mesh_shape = list(mesh_device.shape)
+
         cache = torch.zeros(
             (
-                MLA1D.MAX_BATCH_SIZE,  # TODO: Split batch when addign DP support
+                MLA1D.MAX_BATCH_SIZE // (mesh_shape[1] if mode == "decode" else 1),  # Prefill does not support DP yet
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/mla_1d.py]: 1. MLA is not gonna run DP
2. the TP mesh dim is dim 1, not 0

File: models/demos/deepseek_v3/tests/test_mla_1d.py (hunk 2 above)
Comment (barci2) [reply models/demos/deepseek_v3/tests/test_mla_1d.py]: well if prefill is not finished then we can't use it in the decoder block anyways so why need for the flag? Can't you have a unit test creating its own prefill state for now?

Comment (avoraTT) [reply models/demos/deepseek_v3/tests/test_mla_1d.py]: I'd have to duplicate the create_state function in the MLA module. Adding a just a flag is cleaner for now.

Yeah we can't use it in the decoder block right now, but we should have prefill support soon anyways (wip). Merging this in will let us guard against regressions though

File: models/demos/deepseek_v3/tt/mla_1d.py (hunk 4 above)
Comment (avoraTT) [reply models/demos/deepseek_v3/tt/mla_1d.py]: MLA is still running TP, but it will be DP just around the SDPA portion so the cache is not replicated.

> the TP mesh dim is dim 1, not 0

Could you clarify? Where is dim=0 being used here?

File: models/demos/deepseek_v3/tests/test_mla_1d.py (hunk 2 above)
Comment (barci2) [reply models/demos/deepseek_v3/tests/test_mla_1d.py]: Sure, but could you at least call this param something different (like is_cache_dp) and have the default value be the data parallel one?

File: models/demos/deepseek_v3/tt/mla_1d.py (hunk 4 above)
Comment (barci2) [reply models/demos/deepseek_v3/tt/mla_1d.py]: if the DP dim would be 1 then TP dim would be 0; it's the other way around

Comment (barci2) [reply models/demos/deepseek_v3/tt/mla_1d.py]: The mesh is (4, 8) with a single activation batch being spread along one of the rows, so the whole module should be TP=8(shape[1]). Here you divide the batch size by mesh_shape[1], which is the wrong dim here.

File: models/demos/deepseek_v3/tt/mla_1d.py [hunk 5]
```diff
@@ -806,10 +885,21 @@ def forward_decode(
             kvpe_cache,
             cur_pos_tensor=position_idxs,
             **cfg["flash_mla"],
-        )  #  [1, bsz, num_heads_local, kv_lora_rank]
+        )  #  [1, bsz_local, num_heads, kv_lora_rank]
         ttnn.deallocate(tt_q)
         attn_out = ttnn.to_memory_config(attn_out, **cfg["flash_mla_out_reshard"])
 
+        # FIXME: All-to-All here!! (attn_out)
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/mla_1d.py:896]: Is this FIXME supposed to be here for this PR?

**Review by barci2 — COMMENTED** [review]
Couple of things regarding the existing files:
1. add_weight_config should be a class method, not an in-function instantiated function
2. It would be great if the afformentioned returning of a constructed dict rather than field-by-field assignment would apply to all of `convert_weights`, `prefill_model_config` and `decode_model_config`
3. You don't assert anywhere the size of the mesh device. From what I understand, part of the MLA is executed on a (1,8) submesh device, and part of it on a (4, 8) mesh_device; yet, I only see one mesh_device argument to each of the config functions.
4. Since CCL1D is bound to a device and therefore non-serializable, it should not be an argument to the model config, but rather to the model state

File: models/demos/deepseek_v3/tt/mla_1d.py (hunk 5 above)
Comment (avoraTT) [reply models/demos/deepseek_v3/tt/mla_1d.py:896]: The FIXME is for future reference, as we currently do not have native CCL support. Please see the attached issue in the PR description.

File: models/demos/deepseek_v3/tt/mla_1d.py (hunk 1 above)
Comment (avoraTT) [reply models/demos/deepseek_v3/tt/mla_1d.py]: done

File: models/demos/deepseek_v3/tests/test_mla_1d.py (hunk 2 above)
Comment (avoraTT) [reply models/demos/deepseek_v3/tests/test_mla_1d.py]: changed

**Comment by avoraTT** [comment]
> Couple of things regarding the existing files:
> 
> 1. add_weight_config should be a class method, not an in-function instantiated function
> 2. It would be great if the afformentioned returning of a constructed dict rather than field-by-field assignment would apply to all of `convert_weights`, `prefill_model_config` and `decode_model_config`
> 3. You don't assert anywhere the size of the mesh device. From what I understand, part of the MLA is executed on a (1,8) submesh device, and part of it on a (4, 8) mesh_device; yet, I only see one mesh_device argument to each of the config functions.
> 4. Since CCL1D is bound to a device and therefore non-serializable, it should not be an argument to the model config, but rather to the model state

These are all valid concerns but they are out of the scope for this PR.

A new PR will be opened to better align the MLA module with the updated Deepseek framework 👍

**Review by yieldthought — APPROVED** [review]
🙌
